### DISCUSSION TITLE
Availability of Terminal Error Feature (shutdown terminal reaction)

### DISCUSSION THREAD

**Discussion by nizarsalhaji94** [confluentinc/parallel-consumer#677] [Q&A]
Hi all,

Is a released date is planned for feature / terminal exception ?

I’m interested to use SHUTDOWN Terminal Failure Reatction.

In fact, we need to shutdown (or stop or kill) the PC after max retry.

Thanks

**Comment by nizarsalhaji94** [comment]
@rkolesnev ?

**Reply by rkolesnev** [reply]
Hi @nizarsalhaji94,

Excuse me for delay in replying - i have missed your question.
We do not plan for that feature right now - of course feel free to submit a PR if you are willing to implement it.
In the mean time - you can either indicate that shutdown is requested from the user function back to you main application thread and initiate shutdown from it. 
Or alternatively just trigger shutdown from userFunction directly on a new thread:
For example:
```
        parallelConsumer.poll(recordContexts -> {
            try {
                processRecord(recordContexts); //Try to process
            } catch (Exception e) { 
                if (exhaustedRetries(recordContexts, e)) { //Got failure - max retries reached?
                    CompletableFuture.runAsync(() -> parallelConsumer.closeDontDrainFirst());
                }
            }
        });
```

**Comment by astubbs** [comment]
@nizarsalhaji94 - a shutdown-after-max-retries reaction is being built in the maintained fork, and there is no release date for it yet.

Today, on [v0.6.0.0](https://github.com/astubbs/parallel-consumer/releases/tag/v0.6.0.0), a throw is always a retry and the way to give up on a record is to swallow the exception; to stop the instance after a retry limit, @rkolesnev's pattern still applies - count the attempts in your function and call `CompletableFuture.runAsync(() -> pc.closeDontDrainFirst())`, never close on the worker thread itself.

What replaces that pattern is the fluent API, an open pull request on the fork, not yet merged (astubbs/parallel-consumer#502). A function returns an outcome - succeeded, filtered, park, or **stop** - and a throw is a retry; a record that exhausts its retry limit (ten by default) is parked in place instead of retried forever, and `stop` asks the instance to shut down, which is the reaction you want. It is listed under "What comes next" on the release page and lands in the feature release after 0.6.0.0; a comment on that pull request saying this is your case is the most useful thing you can do to move it up. The dead-letter queue for records that cannot succeed is the step behind it, designed and not built (astubbs/parallel-consumer#149, this tracker's #310).

The coordinates changed with the fork - on Maven Central it is now:

```xml
<dependency>
    <groupId>bz.stub.parallelconsumer</groupId>
    <artifactId>parallel-consumer-core</artifactId>
    <version>0.6.0.0</version>
</dependency>
```

(`parallel-consumer-vertx`, `parallel-consumer-reactor` and `parallel-consumer-mutiny` likewise.) The Java package is `bz.stub.parallelconsumer` too, so imports change with it - one command rewrites them (GNU `sed`; on macOS use `sed -i ''`):

```bash
find . -name '*.java' -exec sed -i 's/io\.confluent\.parallelconsumer/bz.stub.parallelconsumer/g' {} +
```
